### PR TITLE
Refactor watchstore/watchsync metrics

### PR DESCRIPTION
**Pull request by marseel**
Refactor watchstore/watchsync metrics.

### CODE REVIEW

**Review by chancez — APPROVED** [review]

File: clustermesh-apiserver/main.go [hunk 1]
```diff
@@ -142,6 +143,7 @@ type parameters struct {
 	Clientset      k8sClient.Clientset
 	Resources      apiserverK8s.Resources
 	BackendPromise promise.Promise[kvstore.BackendOperations]
+	Factory        store.WatchSyncStoreFactory
```
Comment (giorio94) [inline clustermesh-apiserver/main.go]: Nit. I'd personally rename this to `StoreFactory` (or something similar), as just plain `Factory` would likely conflict in the future. The same also applies to the other occurrences.

File: pkg/clustermesh/cell.go [hunk 2]
```diff
@@ -35,6 +35,7 @@ var Cell = cell.Module(
 	cell.ProvidePrivate(idsMgrProvider),
 
 	cell.Config(common.Config{}),
+	store.Cell,
```
Comment (giorio94) [inline pkg/clustermesh/cell.go]: I wouldn't register the store cell here, as this is a dependency of the clustermesh cell, and not something that the clustermesh cell provides (a cell can only be registered once in the entire hive). To me, it seems more an infrastructural cell that should be registered in `daemon/cmd/cells.go`. The same also applies to the kvstoremesh cell.

File: pkg/clustermesh/clustermesh_test.go [hunk 3]
```diff
@@ -139,6 +141,13 @@ func TestClusterMesh(t *testing.T) {
 	t.Cleanup(func() { ipc.Shutdown() })
 
 	usedIDs := NewClusterMeshUsedIDs()
+	var st store.WatchSyncStoreFactory
+	hive.New(
+		store.Cell,
+		cell.Invoke(func(s store.WatchSyncStoreFactory) {
+			st = s
+		}),
+	).Populate()
```
Comment (giorio94) [inline pkg/clustermesh/clustermesh_test.go]: I'd personally replace this (and the other occurrences) with a method that simply returns a new factory. Constructing a complete hive is pretty verbose, and IMO also a bit confusing for tests.

File: pkg/kvstore/store/watchstore.go [hunk 4]
```diff
@@ -126,8 +129,12 @@ func (rws *restartableWatchStore) Watch(ctx context.Context, backend WatchStoreB
...
 
 	rws.log = rws.log.WithField(logfields.Prefix, prefix)
-	syncedMetric := metrics.KVStoreInitialSyncCompleted.WithLabelValues(
-		kvstore.GetScopeFromKey(prefix), rws.source, "read")
+
+	syncedMetric := metrics.NoOpGauge
+	if rws.syncMetric != nil {
+		syncedMetric = rws.syncMetric.WithLabelValues(
+			kvstore.GetScopeFromKey(prefix), rws.source, "read")
+	}
```
Comment (giorio94) [inline pkg/kvstore/store/watchstore.go]: Is the if statement necessary? AFAIU syncMetric should never be nil.

**Review by giorio94 — CHANGES_REQUESTED** [review]
Changes overall make sense to me.

I've left inline a few comments and nits. Could you please also extend the commit message with a bit of description about the rationale behind the changes?

**Review by giorio94 — APPROVED** [review]
/lgtm

**Review by christarazi — APPROVED** [review]
LGTM from my codeowners

**Comment by marseel** [comment]
rebased onto main branch

**Comment by marseel** [comment]
Removing Casey as he is OOO

**Review by jrajahalme — APPROVED** [review]
LGTM

**Review by ti-mo — APPROVED** [review]

**Pull request merged**